hooks/resolve_publish.py

```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+from sgtk import Hook
+
+
+class PublishResolver(Hook):
+    """
+    Hook used to resolve publish records in Shotgun into
+    local form on a machine.
+    """
+
+    def resolve_path(self, sg_publish_data):
+        """
+        Resolves a Shotgun publish record into a local file on disk.
+        If the method returns None, the default implementation will be used.
```

Comment:
the default logic is quite extensive. I thought about it but i got worried that people would cut n paste the code in that case. putting it in the hook means we give up ownership of the code to some degree - it limites future changes we could push to people. Thinking about the benefits of keeping the logic here, what would those be? The intent of this change is to make it easier for people to be able to for example implement a `perforce://` url if they want (in addition to the stuff that the resolve logic does by default). or take it over completely! Does that make sense?